### PR TITLE
fix(search_family): Fix indexes loading in the FT.SEARCH and Ft.AGGREGATE commands

### PR DESCRIPTION
**Pull request by BagritsevichStepan**
fixes dragonflydb#3782, dragonflydb#3659

### CODE REVIEW

File: src/server/search/search_family.cc [hunk 1]
```diff
@@ -183,6 +183,36 @@ optional<search::Schema> ParseSchemaOrReply(DocIndex::DataType type, CmdArgParse
 #pragma GCC diagnostic pop
 #endif
 
+std::string_view ParseField(CmdArgParser* parser) {
+  std::string_view field = parser->Next();
+  if (field.front() == '@') {
+    field.remove_prefix(1);  // remove leading @ if exists
+  }
+  return field;
```
Comment (dranikpg) [inline src/server/search/search_family.cc:191]: parser fails silently, don't call front() unconditionally

File: src/server/search/search_family.cc [hunk 2]
```diff
@@ -183,6 +183,36 @@ optional<search::Schema> ParseSchemaOrReply(DocIndex::DataType type, CmdArgParse
...
+}
+
+std::optional<std::string_view> ParseFieldWithAtSign(CmdArgParser* parser) {
+  std::string_view field = parser->Next();
+  if (field.front() != '@') {
```
Comment (dranikpg) [inline src/server/search/search_family.cc:196]: same here

File: src/server/search/search_family.cc [hunk 3]
```diff
@@ -183,6 +183,36 @@ optional<search::Schema> ParseSchemaOrReply(DocIndex::DataType type, CmdArgParse
...
+    return std::nullopt;  // if we expect @, but it's not there, return nullopt
+  }
+  field.remove_prefix(1);  // remove leading @
+  return field;
+}
+
+void ParseLoadFields(CmdArgParser* parser, std::optional<SearchFieldsList>& load_fields) {
```
Comment (dranikpg) [inline src/server/search/search_family.cc:203]: before Roman finds it: nit mutable values are passed by pointer 🙂

File: src/server/search/doc_accessors.h [hunk 4]
```diff
@@ -28,14 +28,25 @@ struct BaseAccessor : public search::DocumentAccessor {
...
+
+  // Serialize selected fields
+  SearchDocData Serialize(const search::Schema& schema, const SearchFieldsList& fields) const {
+    SearchFieldsSvList sv_fields;
+    sv_fields.reserve(fields.size());
+    for (const auto& [ident, name] : fields) {
+      sv_fields.emplace_back(ident, name);
+    }
+    return Serialize(schema, sv_fields);
+  }
```
Comment (dranikpg) [inline src/server/search/doc_accessors.h:49]: Not sure why you suddently need both string and string_views 🤷🏻‍♂️ If you really do, you can avoid duplicating the interface by using base::it::Wrap

```c++

template<typename S>
using FieldSpan = absl::Span<const std::pair<S, S>>;

void Serialize(variant<FieldSpan<string>, FieldSpan<string_view>> fields) {
  auto mapper = [](const auto& p) -> pair<string_view, string_view> { return {{p.first}, {p.second}}; };
  for (auto [name, ident] : base::it::Wrap(mapper, fields))
    ...
}

```

File: src/server/search/search_family.cc (hunk 3 above)
Comment (BagritsevichStepan) [reply src/server/search/search_family.cc:203]: Yes, it was temporary

Comment (romange) [reply src/server/search/search_family.cc:203]: 😄

File: src/server/search/doc_accessors.cc [hunk 5]
```diff
@@ -80,7 +80,7 @@ search::SortableValue ExtractSortableValueFromJson(const search::Schema& schema,
 }  // namespace
 
 SearchDocData BaseAccessor::Serialize(const search::Schema& schema,
-                                      const FieldsList& fields) const {
+                                      const SearchFieldsSvList& fields) const {
   SearchDocData out{};
   for (const auto& [fident, fname] : fields) {
     out[fname] = ExtractSortableValue(schema, fident, absl::StrJoin(GetStrings(fident), ","));
```
[the remark this answers is not in the document]
Comment (BagritsevichStepan) [reply src/server/search/doc_accessors.cc:86]: It returns sortable value?

[the remark this answers is not in the document]
Comment (dranikpg) [reply src/server/search/doc_accessors.cc:86]: Yeah, but it's not exclusively used for sorting - that's the confusing part. Its used for all document access. It just returns a document value in one of the value types

**Review by dranikpg — APPROVED** [review]

**Pull request merged**